inference-engine/src/vpu/graph_transformer/src/middleend/passes/gemm_transpose.cpp

```diff
@@ -56,23 +56,36 @@ void PassImpl::run(const Model& model) {
 
         const auto inputDimsA = inputA->desc().dims();
 
-        const auto K = inputDimsA[Dim::H];
-        const auto M = inputDimsA[Dim::W];
-        const auto batch1 = inputDimsA[Dim::N];
-        const auto batch2 = inputDimsA[Dim::C];
+        VPU_THROW_UNLESS(inputDimsA.size() >= 2,
+            "Processing layer {} with type {} failed: first input ({} with usage {}) should have at least 2 dimensions, but it actually has {}",
+            stage->name(), stage->type(), inputA->name(), inputA->usage(), inputDimsA.size());
 
-        const auto inputATranspose = model->duplicateData(inputA, "@reshape", DataDesc{K, M, batch2, batch1});
+        const auto perm = DimsOrder::fromNumDims(inputDimsA.size()).toPermutation();
+
+        std::vector<int> batchDims;
+        DimValues_<Dim> permMap = { {perm[0], perm[1]}, {perm[1], perm[0]} };
+        for (std::size_t i = 2; i < inputDimsA.size(); i++) {
```

Comment:
Why do you start with `2`?

Reply:
According to the MatMul specification (see the link above) the two very last dimensions means rows and columns while all left-most axes (if present) are interpreted as batch dimensions. `DimsOrder::fromNumDims(inputDimsA.size()).toPermutation();` gives us the dimensions in reversed order, which means that in our case the first two dimensions are rows and columns while all dimensions starting from 2 are interpreted as batch dimensions, so in this `for` loop I just process this remaining batch dimensions.

Reply:
Thanks, I got it.